File: src/general/g.opencode/.opencode/tool/grass_g_region.ts
import { tool, question } from "@opencode-ai/plugin"
import { execSync } from "child_process"
import * as fs from "fs"
import * as path from "path"

export default tool({
  description: "Comprehensive GRASS region management: display, set, create regions, and manage boundaries/resolutions.",
  args: {
    gisbase: tool.schema.string().optional().describe("Path to GRASS GIS installation (GISBASE)"),
    gisdbase: tool.schema.string().optional().describe("Path to GRASS database (GISDBASE)"),
    project: tool.schema.string().optional().describe("GRASS project name"),
    mapset: tool.schema.string().optional().describe("GRASS mapset name"),
    flags: tool.schema.string().optional().describe("Flags for g.region command (leave empty for interactive menu)"),
    output_format: tool.schema.string().optional().describe("Output format: raw_text, parsed_dict, or json"),
    create_location: tool.schema.boolean().optional().describe("Whether to create a new location"),
    location_name: tool.schema.string().optional().describe("Name for new location"),
    epsg: tool.schema.number().optional().default(4326).describe("EPSG code for new location"),
    north: tool.schema.number().optional().describe("North bound for region"),
    south: tool.schema.number().optional().describe("South bound for region"),
    east: tool.schema.number().optional().describe("East bound for region"),
    west: tool.schema.number().optional().describe("West bound for region"),
    resolution: tool.schema.number().optional().describe("Resolution for region"),
  },
  async execute(args) {
    // Load config from Python export
    const configCmd = `python3 "${path.resolve('.opencode/tool/export_config.py')}"`
    let config
    try {
      const configOutput = execSync(configCmd, { encoding: 'utf-8', cwd: process.cwd() })
      config = JSON.parse(configOutput)
      if (config.error) {
        throw new Error(config.error)
      }
    } catch (error) {
      throw new Error(`Failed to load config: ${error.message}`)
    }

    // Interactive menu if no flags provided
    let flags = args.flags
    if (!flags) {
      const menuOptions = [
        { label: "Basic region info", description: "Print current region (-p)" },
        { label: "3D region info", description: "Print with 3D settings (-p3)" },
        { label: "Shell script style", description: "Print in key=value format (-g)" },
        { label: "Flat shell style", description: "One-line key=value format (-f)" },
        { label: "Lat/long format", description: "Geographic coordinates (-l)" },
        { label: "Extent only", description: "Print boundaries (-e)" },
        { label: "Center coordinates", description: "Print map center (-c)" },
        { label: "GMT style", description: "GMT format (-t)" },
        { label: "WMS style", description: "WMS format (-w)" },
        { label: "Resolution in meters", description: "Geodesic meters (-m)" },
        { label: "Convergence angle", description: "Grid vs true north (-n)" },
        { label: "Bounding box", description: "Lat/long on WGS84 (-b)" },
      ]
      const answers = await question({
        questions: [{
          question: "Select the type of region information to display:",
          header: "Region Info Menu",
          options: menuOptions,
          multiple: false
        }]
      })
      const selected = answers[0]
      switch (selected) {
        case "Basic region info": flags = "p"; break
        case "3D region info": flags = "p3"; break
        case "Shell script style": flags = "g"; break
        case "Flat shell style": flags = "f"; break
        case "Lat/long format": flags = "l"; break
        case "Extent only": flags = "e"; break
        case "Center coordinates": flags = "c"; break
        case "GMT style": flags = "t"; break
        case "WMS style": flags = "w"; break
        case "Resolution in meters": flags = "m"; break
        case "Convergence angle": flags = "n"; break
        case "Bounding box": flags = "b"; break
        default: flags = "p"
      }
    }

    // Apply defaults
    const gisbase = args.gisbase || config.GISBASE
    const gisdbase = args.gisdbase || config.GRASS_GISDBASE
    const project = args.project || config.GRASS_LOCATION
    const mapset = args.mapset || config.GRASS_MAPSET
    flags = flags || config.GRASS_REGIONS_FLAG
    const output_format = args.output_format || config.GRASS_REGIONS_OUTPUT_FORMAT

    // Validate paths
    if (!fs.existsSync(gisdbase)) {
      throw new Error(`GISDBASE path does not exist: ${gisdbase}`)
    }
    const mapsetPath = `${gisdbase}/${project}/${mapset}`
    if (!fs.existsSync(mapsetPath)) {
      throw new Error(`Mapset path does not exist: ${mapsetPath}. Failing gracefully.`)
    }

    // Construct and run command
    let command: string
    if (args.create_location) {
      if (!args.location_name || !args.north || !args.south || !args.east || !args.west) {
        throw new Error("For creating location, provide location_name, north, south, east, west")
      }
      command = `python3 .opencode/tool/grass_g_region.py create "${gisbase}" "${gisdbase}" "${args.location_name}" "${args.epsg || 4326}" "${args.north}" "${args.south}" "${args.east}" "${args.west}"`
    } else if (args.resolution) {
      // Set region
      if (!args.north || !args.south || !args.east || !args.west) {
        throw new Error("For setting region, provide north, south, east, west, resolution")
      }
      command = `python3 .opencode/tool/grass_g_region.py set "${gisbase}" "${gisdbase}" "${project}" "${mapset}" "${args.north}" "${args.south}" "${args.east}" "${args.west}" "${args.resolution}"`
    } else {
      // Query
      command = `python3 .opencode/tool/grass_g_region.py "${gisbase}" "${gisdbase}" "${project}" "${mapset}" "${flags}" "${output_format}"`
    }
    try {
      const result = execSync(command, { encoding: 'utf-8', cwd: process.cwd() })
      return result.trim()
    } catch (error) {
      throw new Error(`Command failed: ${error.message}`)
    }
  },
})

File: src/general/g.opencode/.opencode/tool/grass_g_region.py
"""

Python script to query GRASS GIS region information using grass.script API.

Copyright (c) 2026 David Sampson

"""

import sys
import os
import platform
import subprocess
import json
import config

def handle_create(gisbase, gisdbase, location_name, epsg, n, s, e, w):
    env = os.environ.copy()
    env['GISBASE'] = gisbase
    env['GISDBASE'] = gisdbase

    # Create location
    cmd = ['g.proj', '-c', f'epsg={epsg}', f'location={location_name}']
    result = subprocess.run(cmd, capture_output=True, text=True, env=env, timeout=30)
    if result.returncode != 0:
        return f"Error creating location: {result.stderr.strip()}"

    # Set region in the new location (PERMANENT mapset is auto-created)
    env['LOCATION_NAME'] = location_name
    env['MAPSET'] = 'PERMANENT'
    cmd2 = ['grass', f'{gisdbase}/{location_name}/PERMANENT', '--exec', 'g.region', f'n={n}', f's={s}', f'e={e}', f'w={w}']
    result2 = subprocess.run(cmd2, capture_output=True, text=True, env=env, timeout=30)
    if result2.returncode != 0:
        return f"Error setting region: {result2.stderr.strip()}"

    return f"Location '{location_name}' created with EPSG {epsg} and bounds n={n} s={s} e={e} w={w}"

def handle_set(gisbase, gisdbase, project, mapset, n, s, e, w, res):
    env = os.environ.copy()
    env['GISBASE'] = gisbase
    env['GISDBASE'] = gisdbase
    env['LOCATION_NAME'] = project
    env['MAPSET'] = mapset

    cmd = ['grass', f'{gisdbase}/{project}/{mapset}', '--exec', 'g.region', f'n={n}', f's={s}', f'e={e}', f'w={w}', f'res={res}']
    result = subprocess.run(cmd, capture_output=True, text=True, env=env, timeout=30)
    if result.returncode != 0:
        return f"Error setting region: {result.stderr.strip()}"

    return f"Region set to n={n} s={s} e={e} w={w} res={res}"

def handle_query(gisbase, gisdbase, project, mapset, flags, output_format):
    # Platform detection
    system = platform.system()

    # GISBASE candidates
    gisbase_candidates = {
        "Linux": ["/usr/lib/grass84", "/usr/local/lib/grass84", "/opt/grass84"],
        "Darwin": ["/Applications/GRASS-8.4.app/Contents/Resources", "/usr/local/grass84"],
        "Windows": ["C:\\OSGeo4W64\\apps\\grass\\grass84", "C:\\OSGeo4W\\apps\\grass\\grass84"],
    }.get(system, [])

    # Find valid GISBASE
    gisbase = gisbase if os.path.exists(gisbase) else None
    if not gisbase:
        for path in gisbase_candidates:
            if os.path.exists(path):
                gisbase = path
                break
    if not gisbase:
        gisbase = config.GISBASE

    # Validate paths
    location_path = os.path.join(gisdbase, project)
    mapset_path = os.path.join(location_path, mapset)
    if not os.path.exists(location_path):
        return f"Error: GRASS location does not exist: {location_path}"
    if not os.path.exists(mapset_path):
        return f"Error: Mapset does not exist: {mapset_path}"

    # Run g.region
    env = os.environ.copy()
    env['GISBASE'] = gisbase
    env['GISDBASE'] = gisdbase
    env['LOCATION_NAME'] = project
    env['MAPSET'] = mapset

    cmd = ['grass', f'{gisdbase}/{project}/{mapset}', '--exec', 'g.region']
    for flag in flags:
        cmd.append(f'-{flag}')

    result = subprocess.run(cmd, capture_output=True, text=True, env=env, timeout=30)
    if result.returncode != 0:
        return f"Error: {result.stderr.strip()}"

    output = result.stdout

    # Format output
    if output_format == "raw_text":
        return output
    elif output_format == "parsed_dict":
        lines = output.strip().split("\n")
        result_dict = {}
        for line in lines:
            if ":" in line:
                key, value = line.split(":", 1)
                result_dict[key.strip()] = value.strip()
        return result_dict
    elif output_format == "json":
        lines = output.strip().split("\n")
        result_dict = {}
        for line in lines:
            if ":" in line:
                key, value = line.split(":", 1)
                result_dict[key.strip()] = value.strip()
        return result_dict  # Will be JSON dumped in main
    else:
        return f"Error: Unknown output format: {output_format}"

def main():
    # Handle help flags - check this first before accessing argv
    if len(sys.argv) == 2 and sys.argv[1] in ['--help', '-h', 'help']:
        print("GRASS Region Management Tool")
        print("Queries, sets, and creates GRASS GIS regions using g.region and g.proj commands")
        print()
        print("Usage:")
        print("  Query: python grass_g_region.py <gisbase> <gisdbase> <project> <mapset> <flags> <output_format>")
        print("  Create: python grass_g_region.py create <gisbase> <gisdbase> <location_name> <epsg> <n> <s> <e> <w>")
        print("  Set: python grass_g_region.py set <gisbase> <gisdbase> <project> <mapset> <n> <s> <e> <w> <res>")
        print()
        print("Arguments:")
        print("  gisbase     - Path to GRASS GIS installation (e.g., /usr/lib/grass84)")
        print("  gisdbase    - Path to GRASS database directory (e.g., /home/user/grassdata)")
        print("  project     - GRASS project/location name (e.g., Sample_Python_Addon)")
        print("  mapset      - GRASS mapset name (e.g., PERMANENT)")
        print("  flags       - Flags for g.region command (e.g., 'p' for print)")
        print("  output_format - Output format: raw_text, parsed_dict, or json")
        print("  location_name - Name for new location")
        print("  epsg        - EPSG code for new location (default 4326)")
        print("  n/s/e/w     - Bounds for region")
        print("  res         - Resolution for region")
        print()
        print("Examples:")
        print("  Query: python grass_g_region.py /usr/lib/grass84 /home/user/grassdata myproject PERMANENT p raw_text")
        print("  Create: python grass_g_region.py create /usr/lib/grass84 /home/user/grassdata Ottawa 4326 46 44 -74 -76")
        print("  Set: python grass_g_region.py set /usr/lib/grass84 /home/user/grassdata myproject PERMANENT 46 44 -74 -76 0.01")
        sys.exit(0)

    # Handle different modes
    if len(sys.argv) >= 2 and sys.argv[1] == "create":
        if len(sys.argv) != 10:
            print("Usage for create: python grass_g_region.py create <gisbase> <gisdbase> <location_name> <epsg> <n> <s> <e> <w>")
            sys.exit(1)
        mode = "create"
        gisbase_arg = sys.argv[2]
        gisdbase_arg = sys.argv[3]
        location_name = sys.argv[4]
        epsg = int(sys.argv[5])
        n = float(sys.argv[6])
        s = float(sys.argv[7])
        e = float(sys.argv[8])
        w = float(sys.argv[9])
    elif len(sys.argv) >= 2 and sys.argv[1] == "set":
        if len(sys.argv) != 11:
            print("Usage for set: python grass_g_region.py set <gisbase> <gisdbase> <project> <mapset> <n> <s> <e> <w> <res>")
            sys.exit(1)
        mode = "set"
        gisbase_arg = sys.argv[2]
        gisdbase_arg = sys.argv[3]
        project_arg = sys.argv[4]
        mapset_arg = sys.argv[5]
        n = float(sys.argv[6])
        s = float(sys.argv[7])
        e = float(sys.argv[8])
        w = float(sys.argv[9])
        res = float(sys.argv[10])
    else:
        # Query mode
        if len(sys.argv) != 7:
            print("Usage for query: python grass_g_region.py <gisbase> <gisdbase> <project> <mapset> <flags> <output_format>")
            print("Use --help for detailed information")
            sys.exit(1)
        mode = "query"
        gisbase_arg = sys.argv[1]
        gisdbase_arg = sys.argv[2]
        project_arg = sys.argv[3]
        mapset_arg = sys.argv[4]
        flags = sys.argv[5]
        output_format = sys.argv[6]

    if mode == "create":
        output = handle_create(gisbase_arg, gisdbase_arg, location_name, epsg, n, s, e, w)
    elif mode == "set":
        output = handle_set(gisbase_arg, gisdbase_arg, project_arg, mapset_arg, n, s, e, w, res)
    else:
        output = handle_query(gisbase_arg, gisdbase_arg, project_arg, mapset_arg, flags, output_format)

    print(json.dumps({"mode": mode, "output": output}))


if __name__ == "__main__":
    main()
